front/lib/api/assistant/jit_actions.ts

```diff
@@ -81,46 +79,55 @@ export async function renderConversationForModelJIT({
   const now = Date.now();
   const messages: ModelMessageTypeMultiActions[] = [];
 
-  // Render loop.
-  // Render all messages and all actions.
+  // Render loop: dender all messages and all actions.
   for (const versions of conversation.content) {
     const m = versions[versions.length - 1];
 
     if (isAgentMessageType(m)) {
       const actions = removeNulls(m.actions);
 
-      // This array is 2D, because we can have multiple calls per agent message (parallel calls).
-
-      const steps = [] as Array<{
-        contents: string[];
-        actions: Array<{
-          call: FunctionCallType;
-          result: FunctionMessageTypeModel;
-        }>;
-      }>;
+      // This is a record of arrays, because we can have multiple calls per agent message (parallel
+      // calls).  Actions all have a step index which indicates how they should be grouped but some
+      // actions injected by `getEmulatedAgentMessageActions` have a step index of `-1`. We
+      // therefore group by index, then order and transform in a 2D array to present to the model.
+      const stepByStepIndex = {} as Record<
```

Comment:
nit: I'd rather use `Map()` insead of Record, unless we are serializing as this is not catched by the typechecking system:

```ts
const aRecord:Record<string, {"k": string}> = {};

aRecord["ruc"].k
```